git.Rmd

```diff
@@ -0,0 +1,151 @@
+---
+output: github_document
+---
+
+This document describes git practices we agree by consensus. It focuses on
+_what_ to do and _why_, and only recommends optional resources about _how_ to do
+it. These practices:
+
+* are aspirational but when they can't be followed it's nice to explain why;
+* document heuristics, as detailed instructions may never cover all scenarios;
+* apply to shared commits (maybe after cleaning commits locally).
+
+## tl;dr
+
+* Shared commits^[A _shared_ commit is a commit that someone other than you
+might access, for example, because you pushed it to a remote repository that
+someone else has permission privileges to access.] should not be changed.
+* Changes in each shared commit should be closely related.
+
+## 1. Preserve shared history
+
+Shared commits should never change. Some Git commands preserve the history, but
+others don't. For example, `git commit --amend` and `git rebase --interactive`
```

Comment:
I think in feature branches this would actually be even more dangerous, as someone is more likely to potentially try to commit/ pull from your working branch. If there is a situation in which someone is likely to have pulled your work and is committing on top of it, you should aim to not alter the history. 

It doesn't mean that you can't clean things up before they are merged (ie you can of course squash/ rebase when the PR is approved/ merged), and it definitely doesn't mean that everything that gets pushed upstream/ origin needs to be super pristine, it just means that the act of pushing sorta means you can't just delete that commit like it never existed. 

Tools like `git revert` allow you to revert a commit by adding the inverse commit on top of it, so as to not alter the history, that would be my recommendation instead of git rebase.

Reply:
I tried my best to rewrite more clearly. The idea of not changing shared-commits applies both to workflows that do and don't use forks. I'm finding it hard to express it succinctly and completely. Maybe I need to complement the text with a demo.

Reply:
so, I guess then my suggestion here would be... "Shared commits should never change." is not nearly specific or precise enough given that many people, myself 100% included, will not know what, if any, actions they intend to take will go against this guidance, i.e. "change a shared commit"